Add Transaction Details notice when a dispute has staged evidence

Jinksi opened the pull request:
Fixes #6966 


> **Note**
> This PR is behind feature flag `_wcpay_feature_dispute_on_transaction_page`


#### Changes proposed in this Pull Request

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

This PR adds a notice to the Transaction Details screen when the dispute associated with the transaction has staged evidence (i.e. the merchant has clicked "Save for later" on the dispute challenge form). This notice reminds the merchant that they have a draft response to the dispute.

![image](https://github.com/Automattic/woocommerce-payments/assets/3147296/99dceaf3-6fa5-4450-8960-5dd221d73b3c)

I've also added a test suite for `DisputeDetails`, a copy of the test suite implemented in the WIP PR #7093.

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Enable feature flag by running `update_option( '_wcpay_feature_dispute_on_transaction_page', '1' );` in [WP Console](https://wordpress.org/plugins/wp-console/) or by modifying your database table `wp_options` directly.
* Create a disputed transaction using the card `4000000000000259` at checkout.
* Find the dispute in Payments → Disputes and click it to view the dispute details screen.
* Click on the Transaction ID to view the Transaction Details screen.
* You should **not** see the staged dispute evidence notice.
* Navigate to the dispute details screen again.
* Click **Challenge**.
* Make a small change to this form and click "Save for Later" to stage the dispute evidence submission.
* Navigate back to the Transaction Details screen.
* You **should now** see the staged dispute evidence notice.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

Jinksi committed:
Add EvidenceDetails type annotations

Jinksi committed:
Add notice if dispute has staged evidence

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `7139` or branch name `add/6966-transaction-details-dispute-challenge-in-progress-notice` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=add/6966-transaction-details-dispute-challenge-in-progress-notice) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 917a04c728a5322bd146057bc4c0d69affcecebc
- Build time: 2023-09-07 02:14:22 UTC

_Note: the build is updated when a new commit is pushed to this PR._

Jinksi committed:
Fix gap between notices when staged notice is rendered

Jinksi committed:
Add N/A changelog entry

Jinksi committed:
Merge branch 'develop' into add/6966-transaction-details-dispute-challenge-in-progress-notice

Jinksi committed:
Correct Dispute interface to expect charge: Charge | string

When the Dispute is nested within a Charge, the charge property is a charge id (ch_) rather than a dispute object.

Jinksi committed:
Add component test for DisputeDetails

Jinksi commented on client/types/disputes.d.ts:
```diff
@@ -51,7 +64,7 @@ export interface Dispute {
 	fileSize?: Record< string, number >;
 	reason: DisputeReason;
-	charge: Charge;
+	charge: Charge | string;
```
When the `Dispute` is nested within a `Charge`, the charge property is a charge id string (`ch_`) rather than a dispute object.

brucealdridge reviewed (approved):
Tested and reviewed. LGTM ✅
